src/charm.py

```diff
@@ -80,11 +85,45 @@ def _on_config_changed(self, _):
         else:
             self.unit.status = WaitingStatus("Waiting for Pebble in workload container")
 
+    def _patch_auth_resources(self):
```

Comment:
```suggestion
    def _patch_auth_resources(self) -> bool:
```

Reply:
It looks like the point of returning `bool` here is that we can skip over errors, and return a flag that says "something failed, but we patched as much as we could". Given that to a sysadmin there's no real difference between that and "i'm raising an Exception at the first thing that failed" in terms of the error requiring manual intervention, I think we should go with something like this:

```python
def _patch_auth_resources(self):
    ...
    with open('auth-stuff.yaml') as f:
        for obj in codes.load_all_yaml(f):
            client.patch(...)

def _on_install(self, event):
    try:
        self.unit.status = MaintenanceStatus("Patching auth configuration")
        self._patch_auth_resources()
        self.unit.status = MaintenanceStatus("Creating CRDs")
        self._create_crds()
        ... etc ...
        self.unit.status = ActiveStatus()
    except ApiError as err:
        logging.error("Applying resources failed.")
        logging.error(traceback.format_exc())
        self.unit.status = BlockedStatus(str(err))
```

That also has the advantage of only setting a status in the event handler, which makes it much easier overall to discover what states a charm can be in.

In a broader sense, I think any function that returns a `bool` should probably be named `is_*` or `has_*` and not do actual work. For better or worse, Python uses exceptions for control flow handling, we probably don't want to fight Python on that.

Also, I wonder if we can make our lives simpler by relying on the Kubernetes API server to guarantee atomicity here, so that we don't have to even worry about "most things worked, but not everything". Looking around, I see [this SO question](https://serverfault.com/q/1047438), which says this:

> All API requests are atomic; an operation either completes entirely or not at all

If we can shove the patching and CRD creation into a single API request (or at least all of the patching into one request, and CRD creation into another), that link makes it seem like we won't have to worry about partial success here. Is that something that's possible with `lightkube`?

Reply:
Updated it in latest PR version. Resolving this.